lib/internal/modules/cjs/loader.js

```diff
@@ -802,7 +803,13 @@ Module._load = function(request, parent, isMain) {
   }
 
   const mod = loadNativeModule(filename, request);
-  if (mod?.canBeRequiredByUsers) return mod.exports;
+  if (mod?.canBeRequiredByUsers) {
+    if (!NativeModule.canBeRequiredWithoutScheme(filename)) {
+      throw new ERR_UNKNOWN_BUILTIN_MODULE(filename);
```

Comment:
It seems to me that instead of throwing an error, it might be better to skip to the user module of the name `test`?

```js
if (mod?.canBeRequiredByUsers && NativeModule.canBeRequiredWithoutScheme(filename)) {
  return mod.exports;
}
```

This would allow for `require('test')` to still be a user module without introducing a new error and maybe solve the problem of incompatibility?

I am also wondering if it wouldn't be faster to add a `canBeRequiredWithoutScheme` property [in the module constructor](https://github.com/nodejs/node/pull/42325/files#diff-b76c7c75bddd3058fb05bee3c05f0ba8d2e26668714aabccdff0158b6ab857e9L204) 

```js
this.canBeRequiredByUsers = !StringPrototypeStartsWith(id, 'internal/');
this.canBeRequiredByUsersWithoutScheme = this.canBeRequiredByUsers && NativeModule.canBeRequiredWithoutScheme(id)
```

then it could be used like:

```js
if (mod?.canBeRequiredByUsersWithoutScheme) return mod.exports
```

Reply:
> It seems to me that instead of throwing an error, it might be better to skip to the user module of the name test?

You're right. I'll send a fix for that.

Reply:
For future reference: https://github.com/nodejs/node/pull/42430